src/devtools/pachdev/actions.go

```diff
@@ -0,0 +1,571 @@
+	// pachClient, and others, This is initialized lazily by K8sManifest() and
+	// should only be accessed through it (even internally, by other methods, in
+	// case it hasn't been initialized yet).
+	k8sManifest string
+
+	// err is set if the overall operation has failed, and indicates the cause.
+	err error
+
+	// oldPachdPodName is the name of the pachd pod that will be replaced
+	// (including pod ID and so on). This is used to determined whether 'pachctl
+	// version' is talking to the old pod or the new pod.
+	oldPachdPodName string
+
+	// oldPachVersion is the version of pachyderm that is being replaced. If
+	// this is the same as 'pachVersion', then we don't need to push a new
+	// image.
+	oldPachVersion string
+}
+
+// Option is the type of optional values passed to NewDeployOp
+type Option func(*deployOp) error
+
+// WithPachVersion is an option for NewDeployOp that sets the version of the
+// Pachyderm instance that will be deployed. The argument can be any released
+// version of Pachyderm (e.g. "2.8.1"), or any of the special strings:
+//   - "local" (indicating that a locally-build Pachyderm image should be
+//     deployed; commonly used on Core)
+//   - "latest" (indicating that the most recent minor release of Pachyderm should
+//     be deployed) (TODO)
+//   - nightly (indicating that the most recent nightly release of Pachyderm
+//     should be deployed)
+func WithPachVersion(v string) Option {
+	return func(op *deployOp) error {
+		op.pachVersion = v
+		return nil
+	}
+}
+
+func WithObjectStorage(v string) Option {
+	// WithPachVersion is an option for NewDeployOp that sets the version of the
+	// Pachyderm instance that will be deployed.
+	return func(op *deployOp) error {
+		op.pachVersion = v
+		return nil
+	}
+}
+
+func NewDeployOp(verbose bool, opts ...Option) (*deployOp, error) {
+	result := &deployOp{
+		pachVersion:   "local",
+		objectStorage: "minio",
+	}
+	for _, op := range opts {
+		if err := op(result); err != nil {
+			return nil, err
+		}
+	}
+
+	return result, nil
+}
+
+// ctx and cancel are captured by result.Ctx(), and will not be initialized
+// if that is never called
+func (op *deployOp) Ctx() context.Context {
+	if op.ctx != nil {
+		return op.ctx
+	}
+	op.ctx, op.cancel = context.WithCancel(context.Background())
+	return op.ctx
+}
+
+func (op *deployOp) Cancel() {
+	if op.cancel != nil {
+		op.cancel()
+	}
+}
+
+// KubeClient is an accessor for op.kubeClient that lazily initializes the
+// field if it's not already set.  All access to op.kubeClient should be through
+// this method, even internally, in case it hasn't been initialized yet.
+func (op *deployOp) KubeClient() *kubernetes.Clientset {
+	if op.kubeClient != nil {
+		return op.kubeClient
+	}
+
+	// Load the Kubernetes configuration from the default location
+	config, err := clientcmd.BuildConfigFromFlags("", clientcmd.RecommendedHomeFile)
+	if err != nil {
+		op.err = errors.Wrap(err, "could not load kubernetes config")
```

Comment:
I am working on writing a Go style guide.  One of the pieces of advice will be to avoid "could not" in error objects.  Since it's an error object, "load kubernetes config" is adequate.

I think this form tends to "stack" better, like "create cluster: load kubernetes config: file /foo/bar: no such file or directory", compared to "could not create cluster: could not load kubernetes config: could not open file /foo/bar: no such file or directory".